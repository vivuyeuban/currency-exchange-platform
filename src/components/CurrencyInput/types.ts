import type { TextFieldProps } from '@material-ui/core';

export type TCurrencyInputProps = Omit<TextFieldProps, 'onChange'> & {
    value?: string | number;
    currency: string;
    onChange?: (value: string) => void;
    precision?: number,
};
