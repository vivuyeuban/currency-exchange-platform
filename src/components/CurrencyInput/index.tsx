import React, { ChangeEvent } from 'react';
import TextField from '@material-ui/core/TextField';
import InputAdornment from '@material-ui/core/InputAdornment';
import FormControl from '@material-ui/core/FormControl';

import {
  toCurrencyString,
  getCurrencySymbol,
  stripNonDigits,
  getProperlyRounded,
} from '../../utils/currencies';
import { DEFAULT_CURRENCY } from '../../constants';

import type { TCurrencyInputProps } from './types';

const DEFAULT_VALUE = '';

const CurrencyInput = React.memo(({
  label,
  value,
  currency = DEFAULT_CURRENCY,
  onChange,
  className,
  precision = 2,
  ...props
}: TCurrencyInputProps) => {
  const [inputVal, setInputVal] = React.useState(DEFAULT_VALUE);
  const [isFocused, setIsFocused] = React.useState(false);

  const onFocus = React.useCallback(() => setIsFocused(true), []);

  React.useEffect(() => {
    if (isFocused) {
      return;
    }
    const newVal = String(value) || DEFAULT_VALUE;
    const rounded = getProperlyRounded(newVal, precision);
    const reservedValue = toCurrencyString({
      amount: rounded,
      precision,
    });
    setInputVal(reservedValue);
  }, [value, precision, isFocused]);

  const internalOnChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { value: eventValue } = event.target;
    const digits = stripNonDigits(eventValue);

    const converted = toCurrencyString({
      amount: eventValue,
      precision,
    });
    setInputVal(converted);
    if (onChange) {
      onChange(getProperlyRounded(digits, precision));
    }
  };

  const onBlur = React.useCallback(() => {
    setIsFocused(false);
    if (!inputVal) {
      return;
    }
    const withPrecision = getProperlyRounded(inputVal);
    const converted = toCurrencyString({
      amount: withPrecision,
      precision,
    });
    setInputVal(converted);
  }, [inputVal, precision]);

  return (
    <FormControl className={className} variant="outlined">
      <TextField
        {...props}
        label={label}
        value={inputVal}
        placeholder="0.00"
        onChange={internalOnChange}
        onBlur={onBlur}
        onFocus={onFocus}
        inputProps={{
          style: { textAlign: 'right' },
        }}
        InputLabelProps={{
          shrink: true,
        }}
        InputProps={{
          startAdornment: <InputAdornment data-testid="currency-sign" position="start">{getCurrencySymbol(currency)}</InputAdornment>,
        }}
        variant="outlined"
      />
    </FormControl>
  );
});

export default CurrencyInput;
