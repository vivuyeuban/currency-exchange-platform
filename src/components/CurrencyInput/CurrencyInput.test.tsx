import React from 'react';
import {
  createEvent, fireEvent, render, screen,
} from '@testing-library/react';
import CurrencyInput from './index';

describe('with necessary props', () => {
  let rendered: ReturnType<typeof render>;
  let inputElement: ReturnType<typeof rendered.getByRole>;
  beforeEach(() => {
    rendered = render(<CurrencyInput
      currency="USD"
    />);
    inputElement = rendered.getByRole('textbox');
  });

  test('renders and matches snapshot', () => {
    expect(rendered.container).toMatchSnapshot();
  });

  test('has currency sign and placeholder', () => {
    const currencySignElelement = rendered.getByTestId('currency-sign');
    const withPlaceholder = screen.getByPlaceholderText('0.00');
    expect(withPlaceholder).toBeInTheDocument();
    expect(currencySignElelement).toHaveTextContent('$');
  });

  test('formats value on the fly, adds precision on blur', () => {
    fireEvent.change(inputElement, { target: { value: '10000000' } });
    expect(inputElement).toHaveValue('10,000,000');

    fireEvent.blur(inputElement);
    expect(inputElement).toHaveValue('10,000,000.00');
  });
});

test('can set value through props', () => {
  const { getByRole } = render(<CurrencyInput
    currency="USD"
    value={10000}
    precision={2}
  />);
  const inputElement = getByRole('textbox');
  expect(inputElement).toHaveValue('10,000.00');
});

test('fires onChange to parent', () => {
  const mockFn = jest.fn();
  const { getByRole } = render(<CurrencyInput
    currency="USD"
    onChange={mockFn}
  />);

  const inputElement = getByRole('textbox');
  const event = createEvent('change', inputElement, { target: { value: '100000000' } });
  fireEvent.change(inputElement, event);

  expect(mockFn).toBeCalledTimes(1);
});
