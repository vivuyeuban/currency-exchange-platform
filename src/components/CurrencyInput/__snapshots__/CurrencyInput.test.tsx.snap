// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with necessary props renders and matches snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root"
  >
    <div
      class="MuiFormControl-root MuiTextField-root"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart"
      >
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionStart"
          data-testid="currency-sign"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            $
          </p>
        </div>
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart"
          placeholder="0.00"
          style="text-align: right;"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          style="padding-left: 8px;"
        >
          <legend
            class="PrivateNotchedOutline-legend-2"
            style="width: 0.01px;"
          >
            <span>
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
